File: LeetCode/Medium/0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        l = 0
        r = 0
        if len(set(s)) <= 1:
            return len(set(s))

        seen = set()
        res = 0

        while r < len(s):
            if s[r] not in seen:
                seen.add(s[r])
                r += 1
                res = max(res , r-l)
            else:
                seen.remove(s[l])
                l+=1
        
        return res
        

File: LeetCode/Medium/0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        int res = 0;
        int l = 0;
        HashMap<Character,Integer> seen = new HashMap<>();

        for (int r = 0 ; r<s.length() ; r++){
            char c = s.charAt(r);
            if (seen.containsKey(c) && seen.get(c) >= l){
                l = seen.get(c)+1;
            }
            res = Math.max(res , (r-l)+1);
            seen.put(c, r);
        }
        return res;
    }
}